### PR TITLE
GSTools covariance model as variogram_model input

### PR DESCRIPTION
**Pull request by MuellerSeb**
With this pull request an interface to [GSTools](https://github.com/GeoStat-Framework/GSTools) is provided. The PyKrige routines ``UniversalKriging``, ``UniversalKriging3D``, ``OrdinaryKriging`` and ``OrdinaryKriging3D`` now allow an instance of the GSTools [CovModel](https://geostat-framework.readthedocs.io/projects/gstools/en/latest/tutorial_02_cov.html) as input for ``variogram_model``. Every information needed will be set internally.

An example was added to the README and to the example folder.

### CODE REVIEW

**Comment by MuellerSeb** [comment]
I went through the CI logs and I think the check-fails are not induced by the pull-request related changes.
Is there something broken in your CI setup or did I miss something?
Cheers,
Sebastian

File: pykrige/ok.py [hunk 1]
```diff
@@ -1,811 +1,850 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-from __future__ import unicode_literals
```
Comment (rth) [inline pykrige/ok.py]: I do not understand what's wrong with the GitHub diff here. These `__future__` import lines have not changed and yet github indicated that they did. Maybe something about the line break character? 

In any case we need to resolve this as otherwise it's quite hard to review.

File: pykrige/ok3d.py [hunk 2]
```diff
@@ -173,6 +175,36 @@ def __init__(self, x, y, z, val, variogram_model='linear',
...
 
+        # set up variogram model and parameters...
+        self.variogram_model = variogram_model
+        self.model = None
+        # check if a GSTools covariance model is given
+        # See: https://github.com/GeoStat-Framework/GSTools
+        if hasattr(self.variogram_model, "pykrige_kwargs"):
+            # save the model in the class
+            self.model = self.variogram_model
+            assert self.model.dim == 3, "GSTools: model dim is not 3"
```
Comment (rth) [inline pykrige/ok3d.py]: Better to `raise ValueError(...)`

File: pykrige/ok3d.py [hunk 3]
```diff
@@ -173,6 +175,36 @@ def __init__(self, x, y, z, val, variogram_model='linear',
                  anisotropy_angle_x=0., anisotropy_angle_y=0.,
                  anisotropy_angle_z=0., verbose=False, enable_plotting=False):
 
+        # set up variogram model and parameters...
+        self.variogram_model = variogram_model
+        self.model = None
+        # check if a GSTools covariance model is given
+        # See: https://github.com/GeoStat-Framework/GSTools
```
Comment (rth) [inline pykrige/ok3d.py]: No need to put the link here

File: pykrige/uk.py [hunk 4]
```diff
@@ -1,1076 +1,1109 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-from __future__ import unicode_literals
```
Comment (rth) [inline pykrige/uk.py]: Same issue with the diff. Maybe it's due to the change in permissions?

File: examples/gstools_covmodel.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+                 [1.9, 0.6, 0.56],
+                 [1.1, 3.2, 0.74],
+                 [3.3, 4.4, 1.47],
+                 [4.7, 3.8, 1.74]])
+
+gridx = np.arange(0.0, 5.5, 0.1)
+gridy = np.arange(0.0, 6.5, 0.1)
+
+# a GSTools based covariance model
+model = Gaussian(dim=2, len_scale=1, anis=0.2, angles=0.5, var=0.5, nugget=0.1)
```
Comment (rth) [inline examples/gstools_covmodel.py]: Maybe call this `covariance_model`

**Comment by MuellerSeb** [comment]
Thanks to @LSchueler I could resolve the diff issue. Now I can care about your suggestions. Sorry for the long delay, but I was a bit feared of the rebase.

File: pykrige/ok.py (hunk 1 above)
Comment (MuellerSeb) [reply pykrige/ok.py]: Rebasing resolved this. Diffs are now clearer.

File: pykrige/ok3d.py (hunk 2 above)
Comment (MuellerSeb) [reply pykrige/ok3d.py]: Done.

File: pykrige/ok3d.py (hunk 3 above)
Comment (MuellerSeb) [reply pykrige/ok3d.py]: Removed.

File: pykrige/uk.py (hunk 4 above)
Comment (MuellerSeb) [reply pykrige/uk.py]: Rebasing resolved this. Diffs are now clearer.

File: examples/gstools_covmodel.py (hunk 5 above)
Comment (MuellerSeb) [reply examples/gstools_covmodel.py]: I used ``cov_model``. I think with the comment above, this is ok.

**Comment by MuellerSeb** [comment]
Ping.

File: pykrige/ok.py [hunk 6]
```diff
@@ -288,29 +302,65 @@ def update_variogram_model(self, variogram_model, variogram_parameters=None,
...
             stretching). Scaling is applied in the y-direction.
-        anisotropy_angle (float, optional): CCW angle (in degrees) by
+        anisotropy_angle : float, optional
+            CCW angle (in degrees) by
             which to rotate coordinate system in order to take into
             account anisotropy. Default is 0 (no rotation).
         """
 
+        # set up variogram model and parameters...
+        self.variogram_model = variogram_model
```
Comment (rth) [inline pykrige/ok.py:338]: Can we factorize this code repeated in all estimators, maybe as,
```py
def _setup_variogram_parameters(self, ..)
```
under pykrige.core. There is already an `_initialize_variogram_model` there. All that could probably be made methods of some base class, but that should probably be done in a separate PR..

File: pykrige/ok3d.py [hunk 7]
```diff
@@ -46,13 +46,15 @@ class OrdinaryKriging3D:
...
-    variogram_model : str, optional
+    variogram_model : str or GSTools CovModel, optional
         Specified which variogram model to use; may be one of the following:
         linear, power, gaussian, spherical, exponential, hole-effect.
         Default is linear variogram model. To utilize a custom variogram model,
         specify 'custom'; you must also provide variogram_parameters and
         variogram_function. Note that the hole-effect model is only technically
         correct for one-dimensional problems.
+        You can also use a GSTools
+        CovModel, see https://github.com/GeoStat-Framework/GSTools
```
Comment (rth) [inline pykrige/ok3d.py]: Use RST link instead of raw url
```rst
`GSTools <https://github.com/GeoStat-Framework/GSTools>`_
```

File: README.rst [hunk 8]
```diff
@@ -31,6 +31,10 @@ point and all grid points. With the 'functional' drift capability, the user may
 of the spatial coordinates that define the drift(s). The package includes a module that contains functions
 that should be useful in working with ASCII grid files (`*.asc`).
 
+PyKrige also works smoothly together with the Covariance Model provided by
+`GSTools <https://github.com/GeoStat-Framework/GSTools>`_ v1.1 and higher.
+Have a look at the section: `GSTools covariance models`_.
```
Comment (rth) [inline README.rst]: I would remove this last sentence, as a link to the documentation of another package before the link to pykrige documentation is confusing. User who need this functionality would be able to find the instructions from other links in this PR.

**Comment by rth** [comment]
Also it looks like CI on master may also need fixing (the CI failures are likely unrelated to this PR). Would you be interested in opening a separate PR to look into it? 

Sorry, this package is very little maintained: personally I'm no longer working on krigging related tasks, and we won't be able to merge this PR as long as CI is not green (at least for Python 3.5+).

**Comment by MuellerSeb** [comment]
@rth I can have a look at the CI and make a separate PR.

**Comment by MuellerSeb** [comment]
> Also it looks like CI on master may also need fixing (the CI failures are likely unrelated to this PR). Would you be interested in opening a separate PR to look into it?

Done.

> Sorry, this package is very little maintained: personally I'm no longer working on krigging related tasks, and we won't be able to merge this PR as long as CI is not green (at least for Python 3.5+).

Again, I could offer help with maintaining this package by integrating it to the [GeoStat-Framework](https://github.com/GeoStat-Framework). Since it's an github organization for geostatics in python, it is easy to share maintainance of packages related to this topic.. at least that is what we are aiming at.
We can discuss about that if you want. Don't want to obtrude myself upon you.

File: pykrige/ok.py (hunk 6 above)
Comment (MuellerSeb) [reply pykrige/ok.py:338]: I just tried that, but it turns out to be more difficult than thought. It should be done in a separate PR refactoring the kriging classes. The ``update_variogram_model`` method should be used here, but it has to be adopted, to work with different dimensions. I will leave it as it is now, if that is ok.

**Comment by MuellerSeb** [comment]
should be ready now.

**Comment by rth** [comment]
I'll try to have a look later this week at the PR.

> Again, I could offer help with maintaining this package by integrating it to the GeoStat-Framework. Since it's an github organization for geostatics in python, it is easy to share maintenance of packages related to this topic.. at least that is what we are aiming at.

Maybe try contacting @bsmurphy by email about it and cc me? At least to see if he could give you write access to this repo. Another solution could be to mark this repo as unmaintained with a link to a maintained version at that org. In any case, personally I won't have much availability to help with the maintenance of this package in the future, and it would be good if some solution could be found at least to merge the occasional PR from contributors (also a PhD is a good time to contribute to OSS :) )

**Comment by MuellerSeb** [comment]
ping.
